File: 3/solutions/code.go
package main

import (
	"fmt"
	"math"
)

func largestPrimeFactor(number uint64) uint64 {
	var largestFactor uint64 = 0

	for number%2 == 0 {
		largestFactor = 2
		number /= 2
	}

	var factor uint64 = 3
	for factor <= uint64(math.Sqrt(float64(number))) {
		if number%factor == 0 {
			largestFactor = factor
			number /= factor
		} else {
			factor += 2
		}
	}

	if number > largestFactor {
		largestFactor = number
	}

	return largestFactor
}

func main() {
	number := uint64(600851475143)
	result := largestPrimeFactor(number)
	fmt.Println(result)
}

File: 3/solutions/code.py
import math

def largest_prime_factor(number):
    largest_factor = 0

    # Divide by 2 to remove all even factors
    while number % 2 == 0:
        largest_factor = 2
        number /= 2

    # Divide by odd factors starting from 3
    factor = 3
    while factor <= math.isqrt(number):
        if number % factor == 0:
            largest_factor = factor
            number /= factor
        else:
            factor += 2

    # If the number is prime, it is the largest prime factor
    if number > largest_factor:
        largest_factor = number

    return largest_factor

number = 600851475143
result = largest_prime_factor(number)
print(result)
